### PR TITLE
Add custom icon support for services

### PR DESCRIPTION
**Pull request by codyfish**
Here is my pull request! As I have never done this before, I hope that I did it correctly.

**What works:**

- Change Icon of service by selecting a custom file using a dialog

**What doesn't work:**

- Changing the icon without a restart
- Syncing the icon with the desktop icon (I would need some settings instance in the application Dialog)

**What would be nice if it worked:**

- Only using file selector to select items, however at the moment you can type default in order to use the icon from gresources.

### CODE REVIEW

File: src/serviceDialog.js [hunk 1]
```diff
@@ -12,8 +12,67 @@
...
+  const openIconChooserDialog = function() {
+    const filter = new Gtk.FileFilter();
+    filter.add_mime_type("image/png");
+
+    const chooser = new Gtk.FileChooserDialog({
+      action: Gtk.FileChooserAction.OPEN,
+      filter: filter,
+      select_multiple: false,
+      //transient_for: this.window,
+      title: "Open",
```
Comment (sonnyp) [inline src/serviceDialog.js]: title should be more descriptive, `title: "Choose an icon"`

File: src/serviceDialog.js [hunk 2]
```diff
@@ -12,8 +12,67 @@
...
+      const active = widget.get_active_iter()[1];
+
+      const text = model.get_value(active, 1);
+
+      const filter = new Gtk.FileFilter();
+      filter.add_mime_type(text);
+
+      chooser.set_filter(filter);
+    });
+    chooser.set_extra_widget(combo);
```
Comment (sonnyp) [inline src/serviceDialog.js]: The combo is not needed, it makes things more complicated for both the code and the user.

File: src/window.js [hunk 3]
```diff
@@ -24,6 +24,7 @@
   // log(imports.gi.Gio.SettingsBackend.get_default());
 
   const { buildHomePage } = imports.homePage;
+  6;
```
Comment (sonnyp) [inline src/window.js]: ? :)

File: src/tab.js [hunk 4]
```diff
@@ -182,7 +208,7 @@
         },
         // https://gjs-docs.gnome.org/webkit240~4.0_api/webkit2.webview#signal-permission-request
         // ["permission-request"](request) {
-        //   log("permission request");
+        //   git ("permission request");
```
Comment (sonnyp) [inline src/tab.js]: ? :)

File: src/serviceManager.js [hunk 5]
```diff
@@ -22,6 +22,15 @@
...
 
+  function register() {
+    const resource = Resource.load(
+      `${pkg.pkgdatadir}/re.sonny.gigagram.services.gresource`
+    );
+    resource._register();
+  }
+
+  register();
+
```
Comment (sonnyp) [inline src/serviceManager.js]: please remove

File: src/serviceDialog.js [hunk 6]
```diff
@@ -87,6 +151,62 @@
...
+    const fileButton = new Gtk.Button({ label: "choose" });
+    grid.attach(fileButton, 3, 3, 1, 1);
+
+    // Bind it to a function that says what to do when the button is clicked
+    fileButton.connect("clicked", () => {
+      const file = openIconChooserDialog();
+      if (file) {
+        iconEntry.text = file;
+      }
+    });
```
Comment (sonnyp) [inline src/serviceDialog.js]: Could you use FileChooserButton instead? No need to reinvent the wheel, FileChooserButton does exactly this.

**Comment by codyfish** [comment]
I'll work on storing the scaled down version of the files. The unnecessary code should be removed, I'll replace FileChooser soon.

File: src/tab.js [hunk 7]
```diff
@@ -160,18 +195,6 @@
...
-    // const webInspector = webView.get_inspector();
-    // webView.connect("load-changed", (self, load_event) => {
-    //   webInspector.show();
-    // });
-
-    // webView.connect("decide-policy", (self, decision, decision_policy) => {
-    //   if (decision_policy === PolicyDecisionType.NEW_WINDOW_ACTION)
-    //     log(decision);
-    //   log(decision_policy);
-    // });
```
Comment (sonnyp) [inline src/tab.js]: please don't remove original comments

Comment (codyfish) [reply src/tab.js]: Yeah, sure I am sorry. I got confused as I did't find this one (I think)

> `         //   git ("permission request"); `
and also didn't remember adding it, so I removed the whole block I guess.

File: src/serviceDialog.js [hunk 8]
```diff
@@ -12,8 +12,53 @@
...
+
+    chooser.add_button("Cancel", Gtk.ResponseType.CANCEL);
+    chooser.add_button("OK", Gtk.ResponseType.OK);
+
+    // This is to add the 'combo' filtering options
+    const store = new Gtk.ListStore();
+    store.set_column_types([GObj.TYPE_STRING, GObj.TYPE_STRING]);
+    store.set(store.append(), [0, 1], ["png", "image/png"]);
+    store.set(store.append(), [0, 1], ["jpg", "image/jpeg"]);
+    store.set(store.append(), [0, 1], ["svg", "image/svg+xml"]);
```
Comment (sonnyp) [inline src/serviceDialog.js]: the store is not needed anymore

**Comment by codyfish** [comment]
Should I acquire the data dir using pkg.pkgdatadir?

**Comment by sonnyp** [comment]
@codyfish use/reuse/extract `const dataPath = build_filenamev([get_user_data_dir(), "gigagram", id]);` that's in `function TabPage` it will give you the directory of the instance/tab "data" folder. `~/.local/share/gigagram/instance-id/`

**Comment by sonnyp** [comment]
Please add yourself to the list of contributors if you wish https://github.com/sonnyp/gigagram/blob/master/src/main.js#L145

**Comment by codyfish** [comment]
Do I have to use my real name or can I also use "codyfish" :D

**Comment by sonnyp** [comment]
You can use any name you wish ;)

I just added a handy observeSetting function in this commit https://github.com/sonnyp/gigagram/commit/1f37cf769f10df1a06b276a4b42de714086370de that you can to replace `get('icon'` + `connect`

**Comment by codyfish** [comment]
Okay tanks!

**Comment by sonnyp** [comment]
@codyfish do you mind if I rebase, cleanup a bit and merge this?

**Comment by codyfish** [comment]
Feel free to do it. I'll just have to pull the changes afterwards, don't I?

**Comment by sonnyp** [comment]
> Feel free to do it. I'll just have to pull the changes afterwards, don't I?

Yes.

```
git checkout master
git pull
```

> I think I finally begin to understand how git works

Congrats, it can be overwhelming at the beginning. Feel free to ask if you have specific issues with gigagram's git.

**Comment by codyfish** [comment]
Okay thanks!

**Comment by sonnyp** [comment]
https://github.com/sonnyp/gigagram/pull/29

**Pull request closed without merging**